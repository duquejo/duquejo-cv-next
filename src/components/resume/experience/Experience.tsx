import { cn } from '@/lib/utils';
import type { ExperienceType } from '@/interfaces/experience';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';

export const Experience = ({
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
}: ExperienceType) => {
  return (
    <ol className="first-of-type:pt-0 pt-5 relative border-s border-gray-200">
      <div
        data-test="detail"
        className={cn(
          'absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400',
          isRecent ? 'bg-primary' : 'bg-secondary',
        )}
      />

      {/*Aligned content*/}
      <div className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-test="time"
          className={cn(
            'lg:inline-flex border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-yellow-400' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          <time className="block mt-1 lg:mt-0 lg:before:content-['-'] lg:before:mx-1">
            {end_date}
          </time>
        </span>
        <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
      </div>
      <Card className="ms-4">
        <CardHeader>
          <CardTitle>{project}</CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        {resume && (
          <CardContent>
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-1 text-sm text-justify">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-sm">{resume}</p>
            )}
          </CardContent>
        )}
        {additional_info && (
          <CardFooter>
            <small className="block font-semibold text-justify">{additional_info}</small>
          </CardFooter>
        )}
      </Card>
    </ol>
  );
};
