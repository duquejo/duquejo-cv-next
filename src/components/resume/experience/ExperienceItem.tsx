import { cn } from '@/lib/utils';
import type { ExperienceType } from '@/interfaces/experience.interface';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { Button } from '@/components/ui/button';
import { Badge } from '@/components/ui/badge';

export const ExperienceItem = ({
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
  screenshot,
  url,
}: ExperienceType) => {
  return (
    <ol className="first-of-type:pt-0 pt-5 relative border-s border-gray-200">
      <div
        data-test="detail"
        className={cn(
          'absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400',
          isRecent ? 'bg-primary' : 'bg-secondary',
        )}
      />

      {/*Aligned content*/}
      <div className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-test="time"
          className={cn(
            'lg:inline-flex border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-yellow-400' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          <time className="block mt-1 lg:mt-0 lg:before:content-['-'] lg:before:mx-1">
            {end_date}
          </time>
        </span>
        <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
      </div>
      <Card className="ms-4">
        <CardHeader>
          <CardTitle>{project}</CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        <CardContent className="flex">
          {resume && (
            <>
              <div
                className={cn(
                  'flex flex-col justify-between items-baseline',
                  screenshot && 'basis-6/12',
                )}
              >
                <div className="flex-auto">
                  <span className="font-semibold text-base">Responsibilities</span>
                  {resume.length > 1 ? (
                    <ol className="list-disc ml-4 text-justify marker:text-secondary">
                      {resume.map((r, i) => (
                        <li key={`resume-${i}`} className="my-1 text-sm text-justify">
                          {r}
                        </li>
                      ))}
                    </ol>
                  ) : (
                    <p className="my-1 text-justify text-sm">{resume}</p>
                  )}
                </div>

                {url && (
                  <Button className="mt-5 flex-shrink" variant="secondary">
                    Project&#39;s link
                  </Button>
                )}
              </div>
            </>
          )}
          {screenshot && (
            <picture className="flex flex-grow flex-col px-4 pt-0">
              <div className="aspect-video rounded-xl bg-muted/50" />
            </picture>
          )}
        </CardContent>
        {additional_info && (
          <CardFooter className="flex flex-col items-start gap-y-2">
            <div className="font-semibold text-base">Skills</div>
            <div className="flex gap-1 flex-wrap">
              {additional_info.map((info) => (
                <Badge variant="secondary" key={info}>
                  {info}
                </Badge>
              ))}
            </div>
          </CardFooter>
        )}
      </Card>
    </ol>
  );
};
