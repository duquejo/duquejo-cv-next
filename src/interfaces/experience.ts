export interface ExperienceType {
  start_date: string;
  end_date: string;
  enterprise: string;
  isRecent: boolean;
  project: string;
  role: string;
  resume?: string[];
  additional_info?: string;
}

export type Sizes = 'small' | 'large' | 'medium';
