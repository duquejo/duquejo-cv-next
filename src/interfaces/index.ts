export * from './experience.interface';
export * from './menu-items.interface';
export * from './main-languages.interface';
export * from './services.interface';
export * from './knowledge.interface';
