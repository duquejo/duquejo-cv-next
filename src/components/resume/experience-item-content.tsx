import { cn } from '@/lib';
import { Button } from '@/components/ui/button';
import { ExperienceType } from '@/interfaces';
import Link from 'next/link';
import Image from 'next/image';

interface Props extends Partial<ExperienceType> {
  contentTitle?: string;
  button?: string;
}

export const ExperienceItemContent = ({
  project = 'Image',
  resume,
  button,
  screenshot,
  url,
  contentTitle,
}: Props) => {
  return (
    <div className="flex flex-col lg:flex-row gap-x-4">
      {resume && (
        <div
          className={cn(
            'flex flex-col order-2 lg:order-1 justify-between items-baseline',
            screenshot && 'basis-8/12',
          )}
        >
          <div className="flex-auto">
            {contentTitle && <span className="font-semibold text-base">{contentTitle}</span>}
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-2 text-xs text-justify leading-normal">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-xs leading-normal">{resume}</p>
            )}
          </div>
          {url && button && (
            <Button
              size="sm"
              className="mt-5 lg:flex-shrink lg:w-auto w-full"
              variant="secondary"
              asChild
            >
              <Link href={url} target="_blank" rel="noopener noreferrer nofollow">
                {button}
              </Link>
            </Button>
          )}
        </div>
      )}
      {screenshot && (
        <picture className="order-1 lg:order-2 flex flex-grow flex-col items-center px-0 mb-5 pt-0 lg:px-4 lg:pt-8 lg:mb-0 lg:max-h-60 lg:max-w-xs overflow-hidden">
          <Image
            src={screenshot}
            width="300"
            height="150"
            loading="lazy"
            alt={project}
            className="rounded object-cover shadow-lg w-auto"
          />
        </picture>
      )}
    </div>
  );
};
