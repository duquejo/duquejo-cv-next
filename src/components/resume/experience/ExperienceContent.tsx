import { cn } from '@/lib/utils';
import { Button } from '@/components/ui/button';
import { ExperienceType } from '@/interfaces';

export const ExperienceContent = ({ resume, screenshot, url }: Partial<ExperienceType>) => {
  return (
    <div className="flex flex-col lg:flex-row">
      {resume && (
        <div
          className={cn(
            'flex flex-col order-2 lg:order-1 justify-between items-baseline',
            screenshot && 'basis-6/12',
          )}
        >
          <div className="flex-auto">
            <span className="font-semibold text-base">Responsibilities</span>
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-1 text-sm text-justify">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-sm">{resume}</p>
            )}
          </div>
          {url && (
            <Button className="mt-5 lg:flex-shrink lg:w-auto w-full" variant="secondary">
              Project&#39;s link
            </Button>
          )}
        </div>
      )}
      {screenshot && (
        <picture className="order-1 lg:order-2 flex flex-grow flex-col px-0 lg:px-4 pt-0 mb-5 lg:mb-0">
          <div className="aspect-video rounded-xl bg-muted/50" />
        </picture>
      )}
    </div>
  );
};
