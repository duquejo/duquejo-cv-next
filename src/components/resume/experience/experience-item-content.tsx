import { cn } from '@/lib/utils';
import { Button } from '@/components/ui/button';
import { ExperienceType } from '@/interfaces';
import Link from 'next/link';
import Image from 'next/image';

interface Props extends Partial<ExperienceType> {
  contentTitle: string;
  button: string;
}

export const ExperienceItemContent = ({
  project = 'Image',
  resume,
  button,
  screenshot,
  url,
  contentTitle,
}: Props) => {
  return (
    <div className="flex flex-col lg:flex-row">
      {resume && (
        <div
          className={cn(
            'flex flex-col order-2 lg:order-1 justify-between items-baseline',
            screenshot && 'basis-6/12',
          )}
        >
          <div className="flex-auto">
            <span className="font-semibold text-base">{contentTitle}</span>
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-1 text-xs text-justify leading-normal">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-xs leading-normal">{resume}</p>
            )}
          </div>
          {url && (
            <Button
              size="sm"
              className="mt-5 lg:flex-shrink lg:w-auto w-full"
              variant="secondary"
              asChild
            >
              <Link href={url} target="_blank" rel="noopener noreferrer nofollow">
                {button}
              </Link>
            </Button>
          )}
        </div>
      )}
      {screenshot && (
        <picture className="order-1 lg:order-2 flex flex-grow flex-col items-center px-0 mb-5 pt-0 lg:px-4 lg:pt-8 lg:mb-0">
          <Image
            src={screenshot}
            width="300"
            height="150"
            loading="lazy"
            alt={project}
            className="rounded object-cover shadow-lg w-auto lg:max-w-xs"
          />
        </picture>
      )}
    </div>
  );
};
