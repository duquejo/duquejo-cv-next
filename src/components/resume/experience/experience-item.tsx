import { cn } from '@/lib/utils';
import type { ExperienceType } from '@/interfaces/experience.interface';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import { ExperienceItemContent } from '@/components/resume/experience/experience-item-content';

export const ExperienceItem = ({
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
  screenshot,
  url,
}: ExperienceType) => {
  return (
    <ol className="first-of-type:pt-0 pt-5 relative border-s border-gray-200">
      <div
        data-testid="detail"
        className={cn(
          'absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400',
          isRecent ? 'bg-primary' : 'bg-secondary',
        )}
      />

      {/*Aligned content*/}
      <div className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-testid="time"
          className={cn(
            'lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-primary' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          {end_date && (
            <time className="block lg:before:content-['-'] lg:before:mx-1">{end_date}</time>
          )}
        </span>
        {enterprise && (
          <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
        )}
      </div>
      <Card className="ms-4">
        <CardHeader>
          <CardTitle>{project}</CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        <CardContent className="py-0">
          <ExperienceItemContent
            project={project}
            resume={resume}
            screenshot={screenshot}
            url={url}
          />
        </CardContent>
        {additional_info && (
          <CardFooter className="flex flex-col items-start gap-y-2 pt-5">
            <div className="font-semibold text-base">Skills</div>
            <div className="flex lg:gap-1 gap-2 flex-wrap lg:justify-normal justify-around">
              {additional_info.map((info) => (
                <Badge variant="secondary" key={info}>
                  {info}
                </Badge>
              ))}
            </div>
          </CardFooter>
        )}
      </Card>
    </ol>
  );
};
