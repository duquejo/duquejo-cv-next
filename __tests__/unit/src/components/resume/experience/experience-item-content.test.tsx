import { ExperienceItemContent } from '@/components/resume/experience/experience-item-content';
import { render, screen } from '@testing-library/react';
import { ExperienceType } from '@/interfaces';

describe('<ExperienceItemContent /> tests', () => {
  const { resume, screenshot, url }: Partial<ExperienceType> = {
    project: 'foo bar',
    resume: ['foo', 'bar', 'baz'],
    screenshot: '/foo/bar.webp',
    url: 'https://foo.bar',
  };

  it('Should match the snapshot with the default args', () => {
    const { container } = render(<ExperienceItemContent />);

    expect(container).toMatchSnapshot();

    expect(screen.queryByRole('list')).toBeNull();
    expect(screen.queryByRole('link')).toBeNull();
    expect(screen.queryByRole('listitem')).toBeNull();
    expect(screen.queryByRole('img')).toBeNull();
  });

  it('Should match the snapshot with the available args - >1 resume items', () => {
    const { container } = render(
      <ExperienceItemContent resume={resume} screenshot={screenshot} url={url} />,
    );

    expect(container).toMatchSnapshot();

    expect(screen.getByRole('list')).toBeInTheDocument();
    expect(screen.getAllByRole('listitem')).toHaveLength(resume.length);

    const cta = screen.getByRole('link');
    expect(cta).toBeInTheDocument();
    expect(cta).toHaveTextContent(/Project's link/);
    expect(cta).toHaveAttribute('href', url);

    const screenShot = screen.getByRole('img');
    expect(screenShot).toBeInTheDocument();
  });

  it('Should render a single paragraph if the resume array is equal to 1.', () => {
    const resumeParagraph = ['baz'];

    render(<ExperienceItemContent resume={resumeParagraph} />);

    expect(screen.queryByRole('list')).toBeNull();
    expect(screen.queryAllByRole('listitem')).toHaveLength(0);

    expect(screen.getByText(resumeParagraph[0])).toBeInTheDocument();
  });
});
