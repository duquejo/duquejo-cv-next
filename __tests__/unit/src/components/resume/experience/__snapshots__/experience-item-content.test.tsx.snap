// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ExperienceItemContent /> tests > Should match the snapshot with the available args - >1 resume items 1`] = `
<div>
  <div
    class="flex flex-col lg:flex-row"
  >
    <div
      class="flex flex-col order-2 lg:order-1 justify-between items-baseline basis-8/12"
    >
      <div
        class="flex-auto"
      >
        <span
          class="font-semibold text-base"
        >
          Mocked title
        </span>
        <ol
          class="list-disc ml-4 text-justify marker:text-secondary"
        >
          <li
            class="my-1 text-xs text-justify leading-normal"
          >
            foo
          </li>
          <li
            class="my-1 text-xs text-justify leading-normal"
          >
            bar
          </li>
          <li
            class="my-1 text-xs text-justify leading-normal"
          >
            baz
          </li>
        </ol>
      </div>
      <a
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-8 rounded-md px-3 text-xs mt-5 lg:flex-shrink lg:w-auto w-full"
        href="https://foo.bar"
        rel="noopener noreferrer nofollow"
        target="_blank"
      >
        Mocked button
      </a>
    </div>
    <picture
      class="order-1 lg:order-2 flex flex-grow flex-col items-center px-0 mb-5 pt-0 lg:px-4 lg:pt-8 lg:mb-0"
    >
      <img
        alt="Image"
        class="rounded object-cover shadow-lg w-auto lg:max-w-xs"
        data-nimg="1"
        decoding="async"
        height="150"
        loading="lazy"
        src="/_next/image?url=%2Ffoo%2Fbar.webp&w=640&q=75"
        srcset="/_next/image?url=%2Ffoo%2Fbar.webp&w=384&q=75 1x, /_next/image?url=%2Ffoo%2Fbar.webp&w=640&q=75 2x"
        style="color: transparent;"
        width="300"
      />
    </picture>
  </div>
</div>
`;

exports[`<ExperienceItemContent /> tests > Should match the snapshot with the default args 1`] = `
<div>
  <div
    class="flex flex-col lg:flex-row"
  />
</div>
`;
